### PR TITLE
TableChangesScan::execute and end to end testing for CDF

### PR DESCRIPTION
**Pull request by OussamaSaoudi-db**
## What changes are proposed in this pull request?
This PR introduces the `execute` method to `TableChangesScan`, which performs a CDF scan and returns `ScanResult` the change data feed. The `ScanResult` holds engine data and a selection vector to apply to the rows of the data.

A helper method `read_scan_file` is added to read the rows of a `scan_file` and perform physical to logical transformation. 

The macros `sort_lines` and `asert_batches_sorted_eq` are moved to the `common` crate so they can  be shared between the `read.rs` and new `cdf.rs` integration tests.

The `cdf-table` and `cdf-table-non-partitioned` test tables are from the delta-rs project.

## How was this change tested?
This tests data reading for 3 tables with the following features:
- a table with deletion vectors involving deletion and restoration
- a non-partitioned table
- a partitioned table

### CODE REVIEW

**Comment by OussamaSaoudi-db** [comment]
Note: This depends on in commit timestamps for the tests to pass. That is incoming.

**Comment by OussamaSaoudi-db** [comment]
TODO; rename table-with-cdf-and-dv to cdf-table-with-dv

File: kernel/src/table_changes/scan.rs [hunk 1]
```diff
@@ -188,6 +187,112 @@ impl TableChangesScan {
...
+    let location = table_root.join(&scan_file.path)?;
+    let file = FileMeta {
+        last_modified: 0,
+        size: 0,
+        location,
+    };
+    let read_result_iter =
+        engine
+            .get_parquet_handler()
+            .read_parquet_files(&[file], read_schema, predicate)?;
```
Comment (zachschuermann) [inline kernel/src/table_changes/scan.rs:275]: (don't have to solve now) I wonder if we should introduce this higher in the iterator pipeline we've built up here?

Oversimplification, but instead of something like:
```rust
scan_files
    .map(|scan_file| read_parquet_files([scan_file])) // parquet batches
    .map(|batch| do_stuff) // return final batches
```
could push it up to be
```rust
read_parquet_files(scan_files)  // parquet batches
    .map(|batch| do_stuff) // return final batches
```

Comment (zachschuermann) [reply kernel/src/table_changes/scan.rs:275]: unsure how much state you'd need to track and associate with each batch (if it makes things too complicated)

**Review by zachschuermann — COMMENTED** [review]
flushing comments, i'll let you rebase now that the other PR landed

**Review by scovich — APPROVED** [review]

File: kernel/src/table_changes/scan.rs [hunk 2]
```diff
@@ -29,8 +33,6 @@ pub struct TableChangesScan {
     predicate: Option<ExpressionRef>,
     // The [`ColumnType`] of all the fields in the `logical_schema`
     all_fields: Vec<ColumnType>,
-    // `true` if any column in the `logical_schema` is a partition column
-    have_partition_cols: bool,
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan.rs:32]: `have_partition_cols` was useful in scan to skip building the physical to logical expression. [Source](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/scan/mod.rs#L482). For CDF, we always have to construct an expression because it changes based on the scan file.

File: kernel/src/table_changes/scan.rs [hunk 3]
```diff
@@ -203,6 +200,99 @@ impl TableChangesScan {
...
+        engine: Arc<dyn Engine>,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<ScanResult>>> {
+        let scan_data = self.scan_data(engine.clone())?;
+        let scan_files = scan_data_to_scan_file(scan_data);
+
+        let global_scan_state = self.global_scan_state();
+        let table_root = self.table_changes.table_root().clone();
+        let all_fields = self.all_fields.clone();
+        let predicate = self.predicate.clone();
+        let dv_engine_ref = engine.clone();
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan.rs:219]: Yeah I was surprised to see that too. I suppose an arc clone is cheap enough given all the processing we're doing.

File: kernel/tests/cdf.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            let record_batch: RecordBatch = data
+                .into_any()
+                .downcast::<ArrowEngineData>()
+                .map_err(|_| Error::engine_data_type("ArrowEngineData".to_string()))?
+                .into();
+            if let Some(mask) = mask {
+                Ok(filter_record_batch(&record_batch, &mask.into())?)
+            } else {
+                Ok(record_batch)
+            }
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi-db) [reply kernel/tests/cdf.rs:49]: I think I'd lean towards the match since it feels clearer that there are only two cases.

File: kernel/src/table_changes/scan.rs [hunk 5]
```diff
@@ -203,6 +200,99 @@ impl TableChangesScan {
...
+        // will cover the following results. we `take()` out of `selection_vector` to avoid
+        // trying to return a captured variable. We're going to reassign `selection_vector`
+        // to `rest` in a moment anyway
+        let mut sv = selection_vector.take();
+        let rest = split_vector(sv.as_mut(), len, None);
+        let result = ScanResult {
+            raw_data: logical,
+            raw_mask: sv,
+        };
+        selection_vector = rest;
```
[the remark this answers is not in the document]
Comment (zachschuermann) [reply kernel/src/table_changes/scan.rs:292]: just created an issue how about we move over to there? https://github.com/delta-io/delta-kernel-rs/issues/587

File: kernel/src/scan/mod.rs [hunk 6]
```diff
@@ -160,7 +160,7 @@ impl ScanResult {
 /// store the name of the column, as that's all that's needed during the actual query. For
 /// `Partition` we store an index into the logical schema for this query since later we need the
 /// data type as well to materialize the partition column.
-#[derive(PartialEq, Debug)]
+#[derive(Clone, PartialEq, Debug)]
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi-db) [reply kernel/src/scan/mod.rs:163]: Ah good point. The reason is really that I don't want to hold a reference to `all_fields` in `TableChangesScan` because the iterator needs to be free from lifetimes.

But really all I need is to `Arc` it. Now we don't need to clone the Vec.

Note that the existing `Scan::execute` still borrows self
```
    pub fn execute(
        &self,
        engine: Arc<dyn Engine>,
    ) -> DeltaResult<impl Iterator<Item = DeltaResult<ScanResult>> + '_> {
```
as you can see with that anonymous lifetime `'_`. I was planning on opening a followup PR to take that out too.

File: kernel/src/table_changes/scan.rs [hunk 7]
```diff
@@ -188,6 +187,112 @@ impl TableChangesScan {
...
+    predicate: Option<ExpressionRef>,
+) -> DeltaResult<impl Iterator<Item = DeltaResult<ScanResult>>> {
+    let ResolvedCdfScanFile {
+        scan_file,
+        mut selection_vector,
+    } = resolved_scan_file;
+
+    let physical_to_logical_expr =
+        physical_to_logical_expr(&scan_file, global_state.logical_schema.as_ref(), all_fields)?;
+    let read_schema = scan_file_read_schema(&scan_file, global_state.read_schema.as_ref());
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan.rs:258]: Ya sure 👍  In that case, we should probably do some renaming in global scan state and `scan/mod.rs` at some point.

File: kernel/src/table_changes/scan.rs (hunk 1 above)
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan.rs:275]: Hmm we'd have to somehow associate the `scan_file` and `selection_vector` to the `batch` because it's important for the P2L transformation.  So the type of the proposed `read_parquet_files` would have to be something like `Iterator<(EngineData, ResolvedCdfScanFile)>`. 
and `do_stuff: (EngineData, ResolvedCdfScanFile) -> ScanResult`

File: kernel/src/table_changes/scan.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan.rs:292]: This took a minute to wrap my head around, but I like it!

File: kernel/src/table_changes/scan.rs [hunk 8]
```diff
@@ -188,6 +187,112 @@ impl TableChangesScan {
...
+        read_schema.clone(),
+        physical_to_logical_expr,
+        global_state.logical_schema.clone().into(),
+    );
+
+    let table_root = Url::parse(&global_state.table_root)?;
+    let location = table_root.join(&scan_file.path)?;
+    let file = FileMeta {
+        last_modified: 0,
+        size: 0,
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan.rs:269]: We discussed this before and the conclusion was that these *may* someday be useful, but immediately they don't contribute much.

iirc, last_modified may be handy because it could be used to check cache staleness in the engine in case it saves a file in memory without re-reading it for each scan.

File: kernel/src/table_changes/scan.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan.rs]: Address this further in this PR too: https://github.com/delta-io/delta-kernel-rs/pull/588/files

File: kernel/src/table_changes/scan.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan.rs:314]: btw ty @zachschuermann  for handling the issue 🙏

File: kernel/src/table_changes/scan.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (scovich) [reply kernel/src/table_changes/scan.rs:240]: Yeah, I don't think there's anything wrong -- it was just the first time I'd seen that and was trying to grok it

File: kernel/src/table_changes/scan.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (zachschuermann) [reply kernel/src/table_changes/scan.rs:290]: right: so I think if the decision is to keep them in filemeta then we should have a follow up to actually use them here (and enable any further optimizations downstream - without this, no matter the downstream optimizations they won't be used)

File: kernel/src/table_changes/scan.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (zachschuermann) [reply kernel/src/table_changes/scan.rs]: cool thanks!

File: kernel/tests/cdf.rs [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+            match mask {
+                Some(mask) => Ok(filter_record_batch(&record_batch, &mask.into())?),
+                None => Ok(record_batch),
+            }
+        })
+        .try_collect()?;
+    Ok(batches)
+}
+
+fn assert_batches_sorted_eq(expected_lines: &[impl ToString], batches: &[RecordBatch]) {
```
Comment (zachschuermann) [inline kernel/tests/cdf.rs:54]: is this duplicated? can we share code or just create follow up to clean this up soon?

**Review by zachschuermann — APPROVED** [review]
LGTM woooo!!! but don't think tests are exhaustive (probably need to test some of the filtering/projection/etc. and i'll need to think through more edge cases) but let's merge and take on some more testing soon!

File: kernel/src/table_changes/scan.rs (hunk 1 above)
Comment (zachschuermann) [reply kernel/src/table_changes/scan.rs]: oh i posted another comment about this, but maybe just make an issue to invest a bit more time into this

File: kernel/tests/cdf.rs (hunk 9 above)
Comment (OussamaSaoudi-db) [reply kernel/tests/cdf.rs]: Ah good catch. Moved it to common and updated tests. We still passing 😎
